tests/test_flow/test_persistence.py

```diff
@@ -747,6 +748,139 @@ def y():  # noqa: F811
     assert f_call_counter.times_called() == 0
 
 
+def test_versioning_auto_with_local_refs(builder, make_counter):
+    call_counter = make_counter()
+
+    builder.set("core__versioning_mode", "auto")
+
+    ret_val = 1
+
+    def ref_func():
+        return ret_val
+
+    @builder
+    @call_counter
+    def x():
+        return ref_func()
+
+    assert builder.build().get("x") == 1
+    assert call_counter.times_called() == 1
+
+    # With no changes, x is not called again.
+    assert builder.build().get("x") == 1
+    assert call_counter.times_called() == 0
+
+    # If we change the ref_func, the bytecode of x changes.
+    def ref_func():  # noqa: F811
+        return ret_val + 7
+
+    assert builder.build().get("x") == 8
+    assert call_counter.times_called() == 1
+
+    # If we change the ret_val, the bytecode of x changes.
+    ret_val = 5
+    assert builder.build().get("x") == 12
+    assert call_counter.times_called() == 1
+
+
+def test_versioning_auto_with_module_refs(builder, make_counter):
+    call_counter = make_counter()
+
+    builder.set("core__versioning_mode", "auto")
+
+    mod_code = """
+    ret_val = 1
+
+    def ref_func():
+        return ret_val
+    """
+    m = import_code(dedent(mod_code))
+
+    @builder
+    @call_counter
+    def x():
+        return m.ref_func()
+
+    assert builder.build().get("x") == 1
+    assert call_counter.times_called() == 1
+
+    # With no changes, x is not called again.
+    m = import_code(dedent(mod_code))
+    assert builder.build().get("x") == 1
+    assert call_counter.times_called() == 0
+
+    # If we change the ref_func, the bytecode of x changes.
+    mod_code = """
+    ret_val = 1
+
+    def ref_func():
+        return ret_val + 7
+    """
+    m = import_code(dedent(mod_code))
+    assert builder.build().get("x") == 8
+    assert call_counter.times_called() == 1
+
+    # If we change the ret_val, the bytecode of x changes.
+    mod_code = """
+    ret_val = 5
+
+    def ref_func():
+        return ret_val + 7
+    """
+    m = import_code(dedent(mod_code))
+    assert builder.build().get("x") == 12
+    assert call_counter.times_called() == 1
+
+
+def test_versioning_auto_version_options(builder):
+    builder.set("core__versioning_mode", "auto")
+
+    complex_number = complex(1.0, 0.0)
+
+    @builder
+    def x():
+        assert complex_number is not None
+        return 1
+
+    with pytest.warns(None):
```

Comment:
What is `warns(None)` doing? My understanding is that this just records the warnings for later use, but we're not using them, are we?

Reply:
Ohhh, I thought it asserts that the warning is `None`. But you are right, it records the warnings. I'll check that there are no recorded warnings with `pytest.warns(None)`.